src/ape_cache/query.py

```diff
@@ -0,0 +1,472 @@
+                check of the number of rows that match the clause.
+
+        Returns:
+            Optional[`sqlalchemy.sql.Expression.Insert`]
+        """
+        pass  # Can't cache this query
+
+    @_cache_update_clause.register
+    def _cache_update_block_clause(self, query: BlockQuery) -> Optional[Insert]:
+        return insert(Blocks)  # type: ignore
+
+    @_cache_update_clause.register
+    def _cache_update_block_txns_clause(self, query: BlockTransactionQuery) -> Optional[Insert]:
+        return insert(Transactions)  # type: ignore
+
+    @_cache_update_clause.register
+    def _cache_update_events_clause(self, query: ContractEventQuery) -> Optional[Insert]:
+        return insert(ContractEvents)  # type: ignore
+
+    @singledispatchmethod
+    def _get_cache_data(
+        self, query: QueryType, result: Iterator[BaseInterfaceModel]
+    ) -> Optional[List[Dict[str, Any]]]:
+        raise QueryEngineError(
+            """
+            Not a compatible QueryType. For more details see our docs
+            https://docs.apeworx.io/ape/stable/methoddocs/exceptions.html#ape.exceptions.QueryEngineError
+            """
+        )
+
+    @_get_cache_data.register
+    def _get_block_cache_data(
+        self, query: BlockQuery, result: Iterator[BaseInterfaceModel]
+    ) -> Optional[List[Dict[str, Any]]]:
+        return [m.dict(by_alias=False) for m in result]
+
+    @_get_cache_data.register
+    def _get_block_txns_data(
+        self, query: BlockTransactionQuery, result: Iterator[BaseInterfaceModel]
+    ) -> Optional[List[Dict[str, Any]]]:
+        new_result = []
+        table_columns = [c.key for c in Transactions.__table__.columns]  # type: ignore
+        for val in [m for m in result]:
+            new_dict = {k: v for k, v in val.dict(by_alias=False).items() if k in table_columns}
+            for col in table_columns:
+                if col == "txn_hash":
+                    new_dict["txn_hash"] = val.txn_hash  # type: ignore
+                elif col == "sender":
+                    new_dict["sender"] = new_dict["sender"].encode()
+                elif col == "receiver" and "receiver" in new_dict:
+                    new_dict["receiver"] = new_dict["receiver"].encode()
+                elif col == "receiver" and "receiver" not in new_dict:
+                    new_dict["receiver"] = bytes()
+                elif col == "block_hash":
+                    new_dict["block_hash"] = query.block_id
+                elif col == "signature":
+                    new_dict["signature"] = val.signature.encode_rsv()  # type: ignore
+                elif col not in new_dict:
+                    new_dict[col] = None
+            new_result.append(new_dict)
+        return new_result
+
+    @_get_cache_data.register
+    def _get_cache_events_data(
+        self, query: ContractEventQuery, result: Iterator[BaseInterfaceModel]
+    ) -> Optional[List[Dict[str, Any]]]:
+        return [m.dict(by_alias=False) for m in result]
```

Comment:
Not exactly happy with all of this, but can fix it later

Reply:
Not my favorite either, but it's required to convert back to the types required for decode_blocks. Might be able to remove this all together with SQLModel

Reply:
Moving this to resolved, will create an issue (or check to make sure we have a SPIKE for SQLModel)